### DISCUSSION TITLE
KiCad database libraries on macOS

### DISCUSSION THREAD

**Discussion by giscus** [cdwilson/cdwilson.dev#6] [Articles]
# KiCad database libraries on macOS

A new “Database Libraries” feature was [recently added](https://gitlab.com/kicad/code/kicad/-/issues/7436) to the KiCad [macOS Nightlies](https://downloads.kicad.org/kicad/macos/explore/nightlies) (6.99) and is scheduled to be officially released in KiCad 7.

In this article, I’m going to show you how to configure KiCad to use a SQLite-based database library with KiCad on macOS.

https://cdwilson.dev/articles/kicad-database-libraries-on-macos/

**Comment by carted01** [comment]
Everything seems to work except for the fact that database parameters that aren't in the actual symbol don't get pulled over from the dblib when placing in the schematic.  For instance, if I drop that capacitor into the schematic, the ${CAPACITANCE} field doesn't resolve because the Capacitance field in the database doesn't get pulled in (even though I see it is there in DB Browser.  Am I maybe missing something?

**Reply by carted01** [reply]
Ok, I think I found the problem.  I believe the qa_dblib.kicad_dbl file doesn't have the capacitor section fully configured for all of the appropriate fields.  The problem doesn't happen with the resistors.

Thanks for this great feature!  It has been a long time coming, and for someone that has been using atomic style components in Altium Designer for years now, I've been dreaming of similar features eventually coming to KiCad!  Bravo!

**Reply by cdwilson** [reply]
@carted01 glad you figured it out. You are correct, the example `qa_dblib.kicad_dbl` is missing the `"Capacitance"` field configuration. I think the devs may just have forgot to put this into the example when they were testing.

For anybody else running across this in the future, here's what you need to add to the `qa_dblib.kicad_dbl` file:

```json
            "fields": [
                ...
                {
                    "column": "Capacitance",
                    "name": "Capacitance",
                    "visible_on_add": false,
                    "visible_in_chooser": false
                },
                ...
            ]
```

After adding the `"Capacitance"` field you should see the Capacitance show up in the symbol chooser:

<img width="366" alt="image" src="https://user-images.githubusercontent.com/14631/208457884-e7950213-7b4f-405c-9979-7efafb4a4142.png">

> Thanks for this great feature!

FYI, I didn't develop this feature (just wrote up the steps here about how to get it working). All thanks go to the KiCad developers for these great new features!

**Reply by carted01** [reply]
Lol... yeah I figured that out... still, much thanks to the graphical breakdown for setting this up in the MacOS version.  Much appreciated!  It is very exciting to see how far this program has evolved over the past decade.

**Comment by FastRabbit** [comment]
Thanks for this tutorial.
Does you know what is the advantage with the redirection for the value with "Resistance" and "Capacitance"?
Would it not possible to have directly the real  vaules in the value filed?

**Reply by cdwilson** [reply]
It would definitely be possible to have the real values in the value field. I think this is just an example showing that it's possible to use the variable substitution in the case where you want to have different fields in each table map to the same "value" field that will show up in KiCad.

**Comment by hraftery** [comment]
Fantastic resource. Despite the arcane sounding steps here, this is indeed the way today and none of it is unnecessary. For example, I was surprised to find "iODBC" still our best option, but found [this reference](https://stackoverflow.com/questions/7548825/what-are-the-functional-differences-between-iodbc-and-unixodbc) useful to understand the players at play.

Perhaps the only superfluous feature is wrapping "SQLite3" in braces in `connection_string`. In my testing it seems to be unnecessary, but doesn't hurt.

Even so, I had a lot of trouble with "Error Loading Library" dialogs at the "Click OK" step, containing the message: "Symbol library 'filename' failed to load. Could not load database library: settings file /path/to/filename.kicad_dbl missing or invalid".

Despite some effort I never managed to produce any logging to help narrow the issue. Instead I started with the known-good example in this article and, after adding/removing lines bit by painful bit, realised at last minute edit meant I'd left an extra comma in my kicad_dbl file.

With a valid kicad_dbl, all was well.

The big hard-earned debug trick I discovered that narrowed the fault finding dramatically was to run this at the command line:

`/Library/Application\ Support/iODBC/bin/iodbctest 'Driver=SQLite3;Database=/path/to/filename.sqlite'`

If that runs, and responds to a command like `tables`, then at least you know the ODBC part is working.

**Reply by cdwilson** [reply]
Thanks for the feedback @hraftery! Glad you found this helpful.
